src/windows.jl

```diff
@@ -25,150 +25,479 @@ export  rect,
+length is the same as an `n+1`-length non-`zerophase` window. Alternatively you
+can think of the continuous `zerophase` window being of width `n` and the
+non-`zerophase` window being of length `n-1`. `zerophase` windows are often used
+in FFT processing, and are usually even-length.
+"""
+
+
+function twoD_docs(arg=nothing)
+    argstr = if arg !== nothing
+        "`$arg`, "
+    else
+        ""
+    end
+    """
+    Providing a 2D `Tuple` rather than a single `n` constructs a 2D window.
+    $argstr`padding` and `zerophase` can then be given either as a single value
+    for both horizontal and vertical or a 2-tuple to specify them separately.
+    """
+end
+
+"""
+    padplot(plotstr)
+
+Takes a multiline string and pre-pads it so that it shows up as
+preformatted code when included in a docstring.
+"""
+function padplot(plotstr)
+    minpad = typemax(Int)
+    lines = split(plotstr, "\n")
+    # count the minimum number of spaces preceeding any line
+    for line in lines
+        line == "" && continue
+        minpad = min(minpad, length(match(r"^ *", line).match))
+    end
+
+    pad = " " ^ (4-minpad)
+    join((string(pad, line) for line in lines), "\n")
+end
+
+include("winplots.jl")
+
+"""
+    makewindow(winfunc::Function, n, padding, zerophase)
+
+Generate a discrete window vector of the given length with `padding` zeros.
+`winfunc` should be a function giving the window value in the range of
+[-0.5, 0.5]. The function is assumed to be 0 outside of this range.
+
+$zerophase_docs
+
+Example:
+
+```
+function hanning(n::Integer; padding=0, zerophase=false)
+    makewindow(n, padding, zerophase) do x
+        0.5*(1+cos(2pi*x))
+    end
+end
+```
+"""
+function makewindow(winfunc::Function, n, padding, zerophase)
+    win = zeros(n+padding)
+    if n == 1
+        win[1] = winfunc(0.0)
+    elseif zerophase
+        # note that the endpoint of the window gets set in both lines. In the
+        # unpadded case this will set the same index (which shouldn't make a
+        # difference if the window is symmetric), but it's necessary for when
+        # there's padding, which ends up in the center of the vector length
+        # n÷2+1
+        win[1:n÷2+1] .= winfunc.(range(0.0, stop=(n÷2)/n, length=n÷2+1))
+        # length n÷2
+        win[end-n÷2+1:end] .= winfunc.(range(-(n÷2)/n, stop=-1/n, length=n÷2))
+    else
+        win[1:n] .= winfunc.(range(-0.5, stop=0.5, length=n))
+    end
+
+    win
+end
+
 #
 # Window functions
 #
 
+# equations rendered into Unicode with https://arthursonzogni.com/Diagon/
+
 """
-    rect(n)
+$rect_winplot
+
+    rect(n; padding=0, zerophase=false)
+    rect((r,c); padding=0, zerophase=false)
+
+Rectangular window of length `n`, padded with `padding` zeros. This window is 1
+within the window, and 0 outside of it.
 
-Rectangular window function of length `n`.
+$(twoD_docs())
+
+$zerophase_docs
 """
-function rect(n::Integer)
-    ones(n)
+function rect(n::Integer; padding=0, zerophase=false)
```

Comment:
It would be nice to have type bounds on KW args to window functions.

Reply:
Agreed. Just pushed a commit that adds tightens the type bounds across the board.